service/matching/taskReader.go

```diff
@@ -421,3 +377,149 @@ func (tr *taskReader) newDispatchContext(isolationGroup string) (context.Context
+			tr.logger.Error("taskReader: unexpected error getting isolation group",
+				tag.Error(err),
+				tag.IsolationGroup(group))
+			tr.completeTask(taskInfo, err)
+			return false, true
+		}
+
+		if group == isolationGroup {
+			// no change, retry to dispatch the task again
+			return false, false
+		}
+
+		// ensure the isolation group is configured and available
+		_, taskGroupReaderIsPresent := tr.taskBuffers[group]
+		if !taskGroupReaderIsPresent {
+			// there's a programmatic error. Something has gone wrong with tasklist instantiation
+			// don't block and redirect to the default group
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectFailureCounter)
+			tr.logger.Error("An isolation group buffer was misconfigured and couldn't be found. Redirecting to default",
+				tag.Dynamic("redirection-from-isolation-group", isolationGroup),
+				tag.Dynamic("redirection-to-isolation-group", group),
+				tag.IsolationGroup(group),
+				tag.WorkflowRunID(taskInfo.RunID),
+				tag.WorkflowID(taskInfo.WorkflowID),
+				tag.TaskID(taskInfo.TaskID),
+				tag.WorkflowDomainID(taskInfo.DomainID),
+			)
+
+			select {
+			case <-tr.cancelCtx.Done():
+				// the task reader is shutting down
+				return true, true
+			case tr.taskBuffers[defaultTaskBufferIsolationGroup] <- taskInfo:
+				// task successfully rerouted to default tasklist
+				return false, true
+			default:
+				// couldn't redirect, loop and try again
+				return false, false
+			}
+		}
+
+		// if there is no poller in the isolation group or the isolation group is drained,
+		// we want to redistribute the tasks to other isolation groups in this case to drain
+		// the backlog.
+		select {
+		case <-tr.cancelCtx.Done():
+			// the task reader is shutting down
+			return true, true
+		case tr.taskBuffers[group] <- taskInfo:
+			// successful redirect
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectCounter)
+			tr.logger.Warn("some tasks were redirected to another isolation group.",
+				tag.Dynamic("redirection-from-isolation-group", isolationGroup),
+				tag.Dynamic("redirection-to-isolation-group", group),
+				tag.WorkflowRunID(taskInfo.RunID),
+				tag.WorkflowID(taskInfo.WorkflowID),
+				tag.TaskID(taskInfo.TaskID),
+				tag.WorkflowDomainID(taskInfo.DomainID),
+			)
+			return false, true
+		default:
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectFailureCounter)
+			tr.logger.Error("some tasks could not be redirected to another isolation group as the buffer's already full",
+				tag.WorkflowRunID(taskInfo.RunID),
+				tag.Dynamic("redirection-from-isolation-group", isolationGroup),
+				tag.Dynamic("redirection-to-isolation-group", group),
+				tag.WorkflowID(taskInfo.WorkflowID),
+				tag.TaskID(taskInfo.TaskID),
+				tag.WorkflowDomainID(taskInfo.DomainID),
+			)
+			// the task async buffers on the other isolation-group are already full, wait and retry
+			return false, false
+		}
+	}
+
+	if errors.Is(err, ErrTasklistThrottled) {
```

Comment:
callout, this is a behaviour change, this previously would just fall into this section for all non-recognised errors and emit the metric for all. I've narrowed the scope that this can happen and put an explicit log around the unrecogised error case which I think should never be valid.

Reply:
useful to separate I think, but we almost definitely want a metric on the unknown-error case.  harder to notice if it's not a metric.

Reply:
true, let me add